### ISSUE TITLE
Getting size mismatch error when trying to train on custom data

### ISSUE THREAD

**Issue by Veboy** [awarebayes/RecNN#14]
Hi,
I've followed the guide for using your own dataset in [this guide](https://recnn.readthedocs.io/en/latest/examples/your_data.html). But I get a bizarre error about shape mismatch.

```
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-fc16f914cab3> in <module>
      6 n_epochs = 2
      7 
----> 8 learn()

<ipython-input-4-9ce1ba07689f> in learn()
      9     for epoch in range(n_epochs):
     10         for batch in env.train_dataloader:
---> 11             loss = ddpg.update(batch, learn=True)
     12             plotter.log_losses(loss)
     13             ddpg.step()

./venv/lib64/python3.8/site-packages/recnn/nn/algo.py in update(self, batch, learn)
     46 
     47     def update(self, batch, learn=True):
---> 48         return self.algorithm(batch, self.params, self.nets, self.optimizers,
     49                               device=self.device, debug=self.debug, writer=self.writer,
     50                               learn=learn, step=self._step)

./venv/lib64/python3.8/site-packages/recnn/nn/update/ddpg.py in ddpg_update(batch, params, nets, optimizer, device, debug, writer, learn, step)
     55     # Value Learning
     56 
---> 57     value_loss = value_update(batch, params, nets, optimizer,
     58                               writer=writer, device=device,
     59                               debug=debug, learn=learn, step=step)

./venv/lib64/python3.8/site-packages/recnn/nn/update/misc.py in value_update(batch, params, nets, optimizer, device, debug, writer, learn, step)
     20 
     21     with torch.no_grad():
---> 22         next_action = nets['target_policy_net'](next_state)
     23         target_value = nets['target_value_net'](next_state, next_action.detach())
     24         expected_value = temporal_difference(reward, done, params['gamma'], target_value)

./venv/lib64/python3.8/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    548             result = self._slow_forward(*input, **kwargs)
    549         else:
--> 550             result = self.forward(*input, **kwargs)
    551         for hook in self._forward_hooks.values():
    552             hook_result = hook(self, input, result)

./venv/lib64/python3.8/site-packages/recnn/nn/models.py in forward(self, state, tanh)
     64         :return: action
     65         """
---> 66         action = F.relu(self.linear1(state))
     67         action = self.drop_layer(action)
     68         action = F.relu(self.linear2(action))

./venv/lib64/python3.8/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    548             result = self._slow_forward(*input, **kwargs)
    549         else:
--> 550             result = self.forward(*input, **kwargs)
    551         for hook in self._forward_hooks.values():
    552             hook_result = hook(self, input, result)

./venv/lib64/python3.8/site-packages/torch/nn/modules/linear.py in forward(self, input)
     85 
     86     def forward(self, input):
---> 87         return F.linear(input, self.weight, self.bias)
     88 
     89     def extra_repr(self):

./venv/lib64/python3.8/site-packages/torch/nn/functional.py in linear(input, weight, bias)
   1608     if input.dim() == 2 and bias is not None:
   1609         # fused op is marginally faster
-> 1610         ret = torch.addmm(bias, input, weight.t())
   1611     else:
   1612         output = input.matmul(weight.t())

RuntimeError: size mismatch, m1: [52039 x 1010], m2: [1290 x 256] at /pytorch/aten/src/TH/generic/THTensorMath.cpp:41
```

This is the network's defition:
```
value_net  = recnn.nn.Critic(1290, 128, 256, 54e-2)
policy_net = recnn.nn.Actor(1290, 128, 256, 6e-1)

cpu_device = torch.device('cpu')
ddpg = recnn.nn.DDPG(policy_net, value_net)
ddpg = ddpg.to(cpu_device)
plotter = recnn.utils.Plotter(ddpg.loss_layout, [['value', 'policy']],)
```

And this is the learning function:
```
# learn function
def learn():
    for epoch in range(n_epochs):
        for batch in env.train_dataloader:
            loss = ddpg.update(batch, learn=True)
            plotter.log_losses(loss)
            ddpg.step()
            if ddpg._step % plot_every == 0:
                clear_output(True)
                print('step', ddpg._step)
                test_loss = run_tests()
                plotter.log_losses(test_loss, test=True)
                plotter.plot_loss()
            if ddpg._step > 100:
                return
```
As you can see, I did almost nothing different to the sample code. So I really have no idea why this shape mismatch would happen. My data also seems identical to ML-20M in terms of structure. For embedding I've used matrix factorization via SVD and each item in it is an array of size 100.

**Comment by awarebayes** [comment]
Hi, if you are using recnn's wrapper, can you please give input shapes of the following:
```
train = env.train_batch()

state, action, reward, next_state, done = recnn.data.get_base_batch(train, device=torch.device('cpu'))
print(state.size(), action.size(), reward.size(), next_state.size(), done.size())
```

**Comment by awarebayes** [comment]
>  I've used matrix factorization via SVD and each item in it is an array of size 100.
See how the networks are defined:
```
value_net  = recnn.nn.Critic(1290, 128, 256, 54e-2)
policy_net = recnn.nn.Actor(1290, 128, 256, 6e-1)
```

From the docs:
recnn.nn.models.Actor(input_dim, action_dim, hidden_size, init_w=0.2)
class recnn.nn.models.Critic(input_dim, action_dim, hidden_size, init_w=3e-05)

Lets recap how we multiply matrices: Let A be a matrix of size [a,b] and B be [c, d]. You can multiply matrices iff b==c. I.E. A [a,b] and B[b, a]. You can only multiply [something, 1010] with [1010, something]

Your action size is 100. State is 10 of these actions + 10 previous rewards. You get 100*10 + 10 = 1010. This is your state size. Change it when you define the networks. 
TL;DR. Specify 
value_net  = recnn.nn.Critic(1010, 100, 256, 54e-2)
policy_net = recnn.nn.Actor(1010, 100, 256, 6e-1)

**Comment by awarebayes** [comment]
P.S. you might need to fiddle with hidden params (init weights, optimizer params, etc..) if you are using it with your own data. I have a full guide how I optimized it for ML 20 M [here](https://towardsdatascience.com/reinforcement-learning-ddpg-and-td3-for-news-recommendation-d3cddec26011?gi=5f65d6f1c45f ) under "Making the DDPG understand" header.

**Comment by Veboy** [comment]
Wow. I feel dumb for not reading the docs lol. Thank you for taking the time to actually look into my issue. You're the best man.
